### PR TITLE
Add pending object to GraphQL queries

### PR DESCRIPTION
**Pull request by shemnon**
## PR description

The initial GraphQL bounty did not include the `pending` object, which was not in the bounty spec.  Add this for completeness.

### CODE REVIEW

File: ethereum/graphqlrpc/src/test/resources/tech/pegasys/pantheon/ethereum/graphqlrpc/graphql_pending.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{
+  "request":
+    "{ pending { transactionCount transactions { from { address } to { address } } account(address:\"0x6295ee1b4f6dd65047762f924ecd367c17eabf8f\") { balance} estimateGas(data:{}) call (data : {from : \"a94f5374fce5edbc8e2a8697c15331677e6ebf0b\", to: \"0x6295ee1b4f6dd65047762f924ecd367c17eabf8f\", data :\"0x12a7b914\"}){data status}} }",
+
+  "response": {
+    "data": {
+      "pending": {
+        "transactionCount": 0,
+        "transactions": [],
```
Comment (mbaxter) [inline ethereum/graphqlrpc/src/test/resources/tech/pegasys/pantheon/ethereum/graphqlrpc/graphql_pending.json]: Can we add some more tests that return non-empty tx's here?

File: ethereum/graphqlrpc/src/main/java/tech/pegasys/pantheon/ethereum/graphqlrpc/internal/TransactionWithMetadata.java [hunk 2]
```diff
@@ -15,37 +15,46 @@
...
+    this.blockHash = Optional.of(blockHash);
+    this.transactionIndex = Optional.of(transactionIndex);
   }
 
   public Transaction getTransaction() {
     return transaction;
   }
 
-  public long getBlockNumber() {
+  public Optional<Long> getBlockNumber() {
```
[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/graphqlrpc/src/main/java/tech/pegasys/pantheon/ethereum/graphqlrpc/internal/TransactionWithMetadata.java]: OptionalLong is not used in this module, so using it in one place has many cascading changes.  More suitable for a cleanup PR than a feature PR.

I opened PIE-1563 to clean these up in general.

File: ethereum/graphqlrpc/src/main/java/tech/pegasys/pantheon/ethereum/graphqlrpc/internal/TransactionWithMetadata.java [hunk 3]
```diff
@@ -15,37 +15,46 @@
...
     return blockNumber;
   }
 
-  public Hash getBlockHash() {
+  public Optional<Hash> getBlockHash() {
     return blockHash;
   }
 
-  public int getTransactionIndex() {
+  public Optional<Integer> getTransactionIndex() {
```
[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/graphqlrpc/src/main/java/tech/pegasys/pantheon/ethereum/graphqlrpc/internal/TransactionWithMetadata.java]: See comment on OptionalLong

File: ethereum/graphqlrpc/src/main/java/tech/pegasys/pantheon/ethereum/graphqlrpc/internal/pojoadapter/PendingStateAdapter.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  // until the miner can expose the current "proposed block" we have no
+  // speculative environment, so estimate against latest.
+  public Optional<AccountAdapter> getAccount(
+      final DataFetchingEnvironment dataFetchingEnvironment) {
+    final BlockchainQuery blockchainQuery =
+        ((GraphQLDataFetcherContext) dataFetchingEnvironment.getContext()).getBlockchainQuery();
+    final Address addr = dataFetchingEnvironment.getArgument("address");
+    final Long bn = dataFetchingEnvironment.getArgument("blockNumber");
+    final long latestBn = blockchainQuery.latestBlock().get().getHeader().getNumber();
+    final Optional<WorldState> ows = blockchainQuery.getWorldState(latestBn);
```
[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/graphqlrpc/src/main/java/tech/pegasys/pantheon/ethereum/graphqlrpc/internal/pojoadapter/PendingStateAdapter.java]: done

File: ethereum/graphqlrpc/src/main/java/tech/pegasys/pantheon/ethereum/graphqlrpc/internal/pojoadapter/PendingStateAdapter.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  // speculative environment, so estimate against latest.
+  public Optional<Long> getEstimateGas(final DataFetchingEnvironment environment) {
+    final Optional<CallResult> result = executeCall(environment);
+    return result.map(CallResult::getGasUsed);
+  }
+
+  // until the miner can expose the current "proposed block" we have no
+  // speculative environment, so estimate against latest.
+  public Optional<CallResult> getCall(final DataFetchingEnvironment environment) {
+    return executeCall(environment);
```
[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/graphqlrpc/src/main/java/tech/pegasys/pantheon/ethereum/graphqlrpc/internal/pojoadapter/PendingStateAdapter.java]: There are two methods specified in the EIP-1767, one that returns the entire call result and one that wants us to estimate the gas.  Executing the call and getting the gasUsed is the easiest way to estimate, so they are deduped by calling executeCall (which is not exposed) and returning what we need.  It could be simplified by making getCall be executeCall.

File: ethereum/graphqlrpc/src/main/java/tech/pegasys/pantheon/ethereum/graphqlrpc/internal/pojoadapter/TransactionAdapter.java [hunk 6]
```diff
@@ -49,18 +48,18 @@ public TransactionAdapter(final TransactionWithMetadata transactionWithMetadata)
...
   }
 
   public Optional<AccountAdapter> getFrom(final DataFetchingEnvironment environment) {
     final BlockchainQuery query = getBlockchainQuery(environment);
-    long blockNumber = transactionWithMetadata.getBlockNumber();
-    final Long bn = environment.getArgument("block");
-    if (bn != null) {
-      blockNumber = bn;
+    final Optional<Long> txBlockNumber = transactionWithMetadata.getBlockNumber();
+    final Optional<Long> bn = Optional.ofNullable(environment.getArgument("block"));
```
[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/graphqlrpc/src/main/java/tech/pegasys/pantheon/ethereum/graphqlrpc/internal/pojoadapter/TransactionAdapter.java]: No, the spec has the `block` param as a Long representing the block number.

File: ethereum/graphqlrpc/src/test/resources/tech/pegasys/pantheon/ethereum/graphqlrpc/graphql_pending.json [hunk 7]
```diff
@@ -0,0 +1,22 @@
+{
+  "request":
+    "{ pending { transactionCount transactions { from { address } to { address } } account(address:\"0x6295ee1b4f6dd65047762f924ecd367c17eabf8f\") { balance} estimateGas(data:{}) call (data : {from : \"a94f5374fce5edbc8e2a8697c15331677e6ebf0b\", to: \"0x6295ee1b4f6dd65047762f924ecd367c17eabf8f\", data :\"0x12a7b914\"}){data status}} }",
```
[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/graphqlrpc/src/test/resources/tech/pegasys/pantheon/ethereum/graphqlrpc/graphql_pending.json]: It's json, which lacks multiline strings.  And the graphql syntax for the call is not json.  So no, not really.

File: ethereum/graphqlrpc/src/test/resources/tech/pegasys/pantheon/ethereum/graphqlrpc/graphql_pending.json (hunk 1 above)
Comment (shemnon) [reply ethereum/graphqlrpc/src/test/resources/tech/pegasys/pantheon/ethereum/graphqlrpc/graphql_pending.json]: mocked in a pending transaction that will return nonce and gas.

File: ethereum/graphqlrpc/src/main/java/tech/pegasys/pantheon/ethereum/graphqlrpc/internal/TransactionWithMetadata.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/graphqlrpc/src/main/java/tech/pegasys/pantheon/ethereum/graphqlrpc/internal/TransactionWithMetadata.java]: One problem is that OptionalLong breaks some Optional<Foo>.map calls.  For example, this won't compile:

```
  public  OptionalLong getEstimateGas(final DataFetchingEnvironment environment) {
    final Optional<CallResult> result = executeCall(environment);
    return result.map(CallResult::getGasUsed);
  }
```
